larpandoracontent/LArMonitoring/TwoViewTransverseTracksValidationTool.h

```diff
@@ -0,0 +1,237 @@
+     *
+     *  @param  t the data to be saved
+     *  @param  datumName the data label
+     *  @param  treeDataBox the box where the data is stored before TTree writing
+     */
+    template <typename T>
+    void StoreAndRegisterDatum(const T &t, const std::string datumName, TreeDataBox &treeDataBox);
+
+    /**
+     *  @brief  Register some data with the pandora monitoring api for writing to a TTree 
+     *
+     *  @param  t the data to be saved
+     *  @param  datumName the data label
+     */
+    template <typename T>
+    void RegisterTreeDatum(T &t, const std::string datumName) const;
+
+    /**
+     *  @brief  Convenience function for collecting and registering truth information with the pandora monitoring api for writing to a TTree 
+     *
+     *  @param  pCluster the cluster whose truth information needs to be stored
+     *  @param  treeDataBox the box where the truth information will be stored
+     *  @param  clusterName a cluster label, needed for TTree branch naming
+     */
+    void CollectTruthInformation(const pandora::Cluster *const pCluster, TreeDataBox &treeDataBox, std::string clusterName);
+
+    /**
+     *  @brief  Convenience function for collecting and registering charge profile information with the pandora monitoring api for writing to a TTree 
+     *
+     *  @param  discreteProbabilityVector the charge profile whose information needs to be stored 
+     *  @param  treeDataBox the box where the charge profile information will be stored
+     *  @param  clusterName a cluster label, needed for TTree branch naming
+     */
+    void CollectChargeProfileInformation(const DiscreteProbabilityVector &discreteProbabilityVector, TreeDataBox &treeDataBox, std::string clusterName);
+
+    /**
+     *  @brief  Check whether two clusters were created by the same true particle 
+     *
+     * @param  pCluster1 the first cluster
+     * @param  pCluster2 the second cluster
+     */
+    bool IsSameMCParticle(const pandora::Cluster *const pCluster1, const pandora::Cluster *const pCluster2);
+
+    pandora::StatusCode ReadSettings(const pandora::TiXmlHandle xmlHandle);
+
+    std::string m_treeName;       ///< name of output tree
+    std::string m_outputFileName; ///< name of output file
+};
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename T>
+void TwoViewTransverseTracksValidationTool::AbstractDatum::emplace_back(const T &t)
+{
+    emplace_back_impl(std::any(t));
+    return;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename T>
+TwoViewTransverseTracksValidationTool::TreeDatum<T>::TreeDatum(
+    const T &t, const TwoViewTransverseTracksValidationTool *const pTool, const std::string datumName) :
+    m_Datum(t)
```

Comment:
So this is the big line, I think. Previously the variable needed to be in scope at the point when FillTree was called. Now that's no longer the case, I think, and indeed can have multiple different algorithms (under same Pandora instance) writing to the same tree until one finally calls FillTree. Allows calls to run off rvalues, tidying things up!

Reply:
Exactly!